### ISSUE TITLE
Discrepency inference test data between dataset

### ISSUE THREAD

**Issue by mgendarme** [MIC-DKFZ/nnUNet#934]
Hello:

I am working on a semantic segmentation task, nuclei segmentation. I am for this playing with two datasets very similar to another (2 different cell lines), the datasets are 2d and got the following task identifiers:
Task006_A549 
Task007_HELA
After training my model I wanted to look at the prediction on the test data (coming from the same distribution but unseen from train - validation):

Command used to train the models (I am using a custom trainer and a custom network architecture):
For sake of simplicity I focus on fold 0 for all the examples.
```
nnUNet_train 2d nnUNetTrainerV2_unet_noDeepSupervision_sn_adam_8e4 Task006_A549 0
nnUNet_train 2d nnUNetTrainerV2_unet_noDeepSupervision_sn_adam_8e4 Task007_HELA 0
```

For Task006_A549
Validation dice = 0.9288 (from postprocessing.json)
I ran this for predicting the test cases:
```
nnUNet_predict -i $nnUNet_raw_data_base/nnUNet_raw_data/Task006_A549/imagesTs -o $RESULTS_FOLDER/nnUNet/2d/Task006_A549/nnUNetTrainerV2_unet_noDeepSupervision_sn_adam_8e4__nnUNetPlansv2.1/imagesTs_predicted_0_blk -t 006 -m 2d -f 0 -tr nnUNetTrainerV2_unet_noDeepSupervision_sn_adam_8e4
nnUNet_evaluate_folder -ref $nnUNet_raw_data_base/nnUNet_raw_data/Task006_A549/labelsTs -pred $RESULTS_FOLDER/nnUNet/2d/Task006_A549/nnUNetTrainerV2_unet_noDeepSupervision_sn_adam_8e4__nnUNetPlansv2.1/imagesTs_predicted_0_blk -l 1
```
Test dice = 0.9318 very close to the validation dice, everything perfect!

Now things get more complicated
For Task007_HELA
Validation dice = 0.9211 (from postprocessing.json)
I ran this for predicting the test cases:
```
nnUNet_predict -i $nnUNet_raw_data_base/nnUNet_raw_data/Task007_HELA/imagesTs -o $RESULTS_FOLDER/nnUNet/2d/Task007_HELA/nnUNetTrainerV2_unet_noDeepSupervision_sn_adam_8e4__nnUNetPlansv2.1/imagesTs_predicted_0_blk -t 007 -m 2d -f 0 -tr nnUNetTrainerV2_unet_noDeepSupervision_sn_adam_8e4
nnUNet_evaluate_folder -ref $nnUNet_raw_data_base/nnUNet_raw_data/Task007_HELA/labelsTs -pred $RESULTS_FOLDER/nnUNet/2d/Task007_HELA/nnUNetTrainerV2_unet_noDeepSupervision_sn_adam_8e4__nnUNetPlansv2.1/imagesTs_predicted_0_blk -l 1
```
Test dice = 0.0166 this is very far from validation dice and I cannot understand why.
If anything I would expect the test dice to be close from validation dice (a bit lower, a bit higher), but here it looks like it is completely off.
I am looking for a way to solve this issue and get correct predictions on the test data.
I have checked and the labels are fine, the model is simply not predicting anything (in the predicted folder the .nii.gz images are almost always empty except some rare cases).
Could it be a problem while loading / saving the model?
Any other clew why that would be? 

Thanks a lot for your help!

----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Additional Testing:

Predicting training data again with saved model for Task007:
```
nnUNet_predict -i $nnUNet_raw_data_base/nnUNet_raw_data/Task007_HELA/imagesTr -o $RESULTS_FOLDER/nnUNet/2d/Task007_HELA/nnUNetTrainerV2_unet_noDeepSupervision_sn_adam_8e4__nnUNetPlansv2.1/imagesTs_predicted_0_blk_tr -t 007 -m 2d -f 0 -tr nnUNetTrainerV2_unet_noDeepSupervision_sn_adam_8e4
nnUNet_evaluate_folder -ref $nnUNet_raw_data_base/nnUNet_raw_data/Task007_HELA/labelsTr -pred $RESULTS_FOLDER/nnUNet/2d/Task007_HELA/nnUNetTrainerV2_unet_noDeepSupervision_sn_adam_8e4__nnUNetPlansv2.1/imagesTs_predicted_0_blk_tr -l 1
```
Validation dice = 0.9211 (from postprocessing.json)
Training dice from inference = 0.0097 clearly something is wrong here !!!

I have done some additional testing:
For Task007_HELA trying to predict using the model trained on Task006_A549 (this has no meaning it is just to identify the problem)
```
nnUNet_predict -i $nnUNet_raw_data_base/nnUNet_raw_data/Task007_HELA/imagesTs -o $RESULTS_FOLDER/nnUNet/2d/Task007_HELA/nnUNetTrainerV2_unet_noDeepSupervision_sn_adam_8e4__nnUNetPlansv2.1/imagesTs_predicted_0_blk_A549 -t 006 -m 2d -f 0 -tr nnUNetTrainerV2_unet_noDeepSupervision_sn_adam_8e4
nnUNet_evaluate_folder -ref $nnUNet_raw_data_base/nnUNet_raw_data/Task007_HELA/labelsTs -pred $RESULTS_FOLDER/nnUNet/2d/Task007_HELA/nnUNetTrainerV2_unet_noDeepSupervision_sn_adam_8e4__nnUNetPlansv2.1/imagesTs_predicted_0_blk_A549 -l 1
```
Test dice = 0.8139 lower than test dice on the data used to trained the model. The model performs worse on data from another distribution but isn't too far off

For Task006_A549 trying to predict using the model trained on Task007_HELA  
```
# nnUNet_predict -i $nnUNet_raw_data_base/nnUNet_raw_data/Task006_A549/imagesTs -o $RESULTS_FOLDER/nnUNet/2d/Task006_A549/nnUNetTrainerV2_unet_noDeepSupervision_sn_adam_8e4__nnUNetPlansv2.1/imagesTs_predicted_0_blk_HELA -t 007 -m 2d -f 0 -tr nnUNetTrainerV2_unet_noDeepSupervision_sn_adam_8e4
# nnUNet_evaluate_folder -ref $nnUNet_raw_data_base/nnUNet_raw_data/Task006_A549/labelsTs -pred $RESULTS_FOLDER/nnUNet/2d/Task006_A549/nnUNetTrainerV2_unet_noDeepSupervision_sn_adam_8e4__nnUNetPlansv2.1/imagesTs_predicted_0_blk_HELA -l 1
```
Test dice = 0.0020 now we are close to nothingness

**Comment by FabianIsensee** [comment]
Hey, it looks like something might be off with the test data. But it's really hard to tell without looking at the dataset. You cannot share it can you?

**Comment by mgendarme** [comment]
Hi Fabian,
Thank you for the answer.
Unfortunately I cannot share the data like this. But I can do any testing you might need on them.

Also I am not sure it is just linked to the test data because when I try to predict the training set again with the saved model:

Predicting training data from Task007 again with model trained on Task007:
```
nnUNet_predict -i $nnUNet_raw_data_base/nnUNet_raw_data/Task007_HELA/imagesTr -o $RESULTS_FOLDER/nnUNet/2d/Task007_HELA/nnUNetTrainerV2_unet_noDeepSupervision_sn_adam_8e4__nnUNetPlansv2.1/imagesTs_predicted_0_blk_tr -t 007 -m 2d -f 0 -tr nnUNetTrainerV2_unet_noDeepSupervision_sn_adam_8e4
nnUNet_evaluate_folder -ref $nnUNet_raw_data_base/nnUNet_raw_data/Task007_HELA/labelsTr -pred $RESULTS_FOLDER/nnUNet/2d/Task007_HELA/nnUNetTrainerV2_unet_noDeepSupervision_sn_adam_8e4__nnUNetPlansv2.1/imagesTs_predicted_0_blk_tr -l 1
```
Validation dice = 0.9211 (from postprocessing.json)
Training dice from inference = 0.0097 clearly something is wrong here !!!
I cannot explaing this the two values should be extremely close.

**Comment by mgendarme** [comment]
I have tried again predicting on train and test images to compare on both data sets:


For Task007_HELA (problemaatic one)
```
echo "Prediction - Task007_HELA - individual fold - fold 0 - Train"
nnUNet_predict -i $nnUNet_raw_data_base/nnUNet_raw_data/Task007_HELA/imagesTr -o $RESULTS_FOLDER/nnUNet/2d/Task007_HELA/nnUNetTrainerV2_unet_noDeepSupervision_sn_adam_8e4__nnUNetPlansv2.1/imagesTr_predicted_fold_0 -t 007 -m 2d -f 0 -tr nnUNetTrainerV2_unet_noDeepSupervision_sn_adam_8e4
nnUNet_evaluate_folder -ref $nnUNet_raw_data_base/nnUNet_raw_data/Task007_HELA/labelsTr -pred $RESULTS_FOLDER/nnUNet/2d/Task007_HELA/nnUNetTrainerV2_unet_noDeepSupervision_sn_adam_8e4__nnUNetPlansv2.1/imagesTr_predicted_fold_0 -l 1
```
Train Dice = 0.0097 | Validation Dice from postprocessing.json = 0.9211
```
echo "Prediction - Task007_HELA - individual fold - fold 0 - Test"
nnUNet_predict -i $nnUNet_raw_data_base/nnUNet_raw_data/Task007_HELA/imagesTs -o $RESULTS_FOLDER/nnUNet/2d/Task007_HELA/nnUNetTrainerV2_unet_noDeepSupervision_sn_adam_8e4__nnUNetPlansv2.1/imagesTs_predicted_fold_0 -t 007 -m 2d -f 0 -tr nnUNetTrainerV2_unet_noDeepSupervision_sn_adam_8e4
nnUNet_evaluate_folder -ref $nnUNet_raw_data_base/nnUNet_raw_data/Task007_HELA/labelsTs -pred $RESULTS_FOLDER/nnUNet/2d/Task007_HELA/nnUNetTrainerV2_unet_noDeepSupervision_sn_adam_8e4__nnUNetPlansv2.1/imagesTs_predicted_fold_0 -l 1
```
Test dice = 0.0166 | Validation Dice from postprocessing.json = 0.9211

For Task006_A549 (the one that works as expected)
```
echo "Prediction - Task006_A549 - individual fold - fold 0 - Train"
nnUNet_predict -i $nnUNet_raw_data_base/nnUNet_raw_data/Task006_A549/imagesTr -o $RESULTS_FOLDER/nnUNet/2d/Task006_A549/nnUNetTrainerV2_unet_noDeepSupervision_sn_adam_8e4__nnUNetPlansv2.1/imagesTr_predicted_fold_0 -t 006 -m 2d -f 0 -tr nnUNetTrainerV2_unet_noDeepSupervision_sn_adam_8e4
nnUNet_evaluate_folder -ref $nnUNet_raw_data_base/nnUNet_raw_data/Task006_A549/labelsTr -pred $RESULTS_FOLDER/nnUNet/2d/Task006_A549/nnUNetTrainerV2_unet_noDeepSupervision_sn_adam_8e4__nnUNetPlansv2.1/imagesTr_predicted_fold_0 -l 1
```
Train Dice = 0.9816 | Validation Dice from postprocessing.json = 0.9288
```
echo " Prediction - Task006_A549 - individual fold - fold 0 - Test"
nnUNet_predict -i $nnUNet_raw_data_base/nnUNet_raw_data/Task006_A549/imagesTs -o $RESULTS_FOLDER/nnUNet/2d/Task006_A549/nnUNetTrainerV2_unet_noDeepSupervision_sn_adam_8e4__nnUNetPlansv2.1/imagesTs_predicted_fold_0 -t 006 -m 2d -f 0 -tr nnUNetTrainerV2_unet_noDeepSupervision_sn_adam_8e4
nnUNet_evaluate_folder -ref $nnUNet_raw_data_base/nnUNet_raw_data/Task006_A549/labelsTs -pred $RESULTS_FOLDER/nnUNet/2d/Task006_A549/nnUNetTrainerV2_unet_noDeepSupervision_sn_adam_8e4__nnUNetPlansv2.1/imagesTs_predicted_fold_0 -l 1
```
Test dice = 0.9318 | Validation Dice from postprocessing.json = 0.9288

**Comment by mgendarme** [comment]
I checked the predictions in more details for Task006_A549
From left to right: ground truth, prediction from validation_raw, prediction done again on training set with same model from same fold
![Montage_A549_015](https://user-images.githubusercontent.com/22475646/152528190-255094b1-8c60-48a3-a408-71e0184862b4.jpg)

I did the same with the predictions for Task007_HELA
From left to right: ground truth, prediction from validation_raw, prediction done again on training set with same model from same fold
![Montage_HeLa_068](https://user-images.githubusercontent.com/22475646/152528146-598d587c-df9c-441c-9d31-aa02ee744fbf.jpg)
It is clear that the model loaded in the case of Task007 has an issue as at then end of training predictions were fine.
I still don't know how to solve it but it helps to delineate where it might come from.

**Comment by FabianIsensee** [comment]
Can you please try this?

https://github.com/MIC-DKFZ/nnUNet/issues/911#issuecomment-1012456391

(you will need to create a new dataset, train again and then test again. You can use a short training to speed things up)
Best,
Fabian

**Comment by FabianIsensee** [comment]
ignore the comment above, that's not your problem. See my email :-)

**Comment by mgendarme** [comment]
> Can you please try this?
> 
> [#911 (comment)](https://github.com/MIC-DKFZ/nnUNet/issues/911#issuecomment-1012456391)
> 
> (you will need to create a new dataset, train again and then test again. You can use a short training to speed things up) Best, Fabian

You might have been right. I have generated a fresh dataset, preprocessed it again (so same procedure as the initial dataset generating issue, nothing was changed) and this time prediction on test set worked just fine. 
This has to be taken with a grain of salt. I have only tested after a couple of epochs. I am gonna run a full training and see if it got rid of the weird behaviour for good.
In any case thanks a lot for you help! I will update my post once training is done and close the issue if the behaviour doesn't re-appear.
